x/feedist/handler.go

```diff
@@ -0,0 +1,300 @@
+	// revenue with no funds can be updated, so that recipients trust us.
+	// Otherwise an admin could change who receives the money without the
+	// previously selected recepients ever being paid.
+	if err := distribute(db, h.ctrl, RevenueAccount(msg.RevenueID), rev.Recipients); err != nil {
+		return res, errors.Wrap(err, "cannot distribute")
+	}
+
+	rev.Recipients = msg.Recipients
+	obj := orm.NewSimpleObj(msg.RevenueID, rev)
+	if err := h.bucket.Save(db, obj); err != nil {
+		return res, errors.Wrap(err, "cannot save")
+	}
+	return res, nil
+}
+
+func (h *updateRevenueHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*UpdateRevenueMsg, error) {
+	rmsg, err := tx.GetMsg()
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get message")
+	}
+	msg, ok := rmsg.(*UpdateRevenueMsg)
+	if !ok {
+		return nil, errors.InvalidMsgErr.New("unknown transaction type")
+	}
+	if err := msg.Validate(); err != nil {
+		return msg, err
+	}
+	return msg, nil
+}
+
+// distribute split the funds stored under the revenue address and distribute
+// them according to recipients proportions. When successful, revenue account
+// has no funds left after this call.
+//
+// It might be that not all funds can be distributed equally. Because of that a
+// small leftover can remain on the revenue account after this operation.
+func distribute(db weave.KVStore, ctrl CashController, source weave.Address, recipients []*Recipient) error {
+	var chunks int64
+	for _, r := range recipients {
+		chunks += int64(r.Weight)
+	}
+
+	// Find the greatest common division for all weights. This is needed to
+	// avoid leaving big fund leftovers on the source account when
+	// distributing between many recipients. Or when there is only one
+	// recipient with a high weight value.
+	var weights []int32
+	for _, r := range recipients {
+		weights = append(weights, r.Weight)
+	}
+	div := findGcd(weights...)
+
+	chunks = chunks / int64(div)
+
+	balance, err := ctrl.Balance(db, source)
+	switch {
+	case err == nil:
+		// All good.
+	case errors.Is(errors.NotFoundErr, err):
+		// Account does not exist, so there is are no funds to split.
+		return nil
+	default:
+		return errors.Wrap(err, "cannot acquire revenue account balance")
+	}
+
+	// TODO normalize balance. There is no functionality that allows to
+	// normalize x.Coins right now (14 Feb 2019).
+
+	// For each currency, distribute the coins equally to the weight of
+	// each recipient. This can leave small amount of coins on the original
+	// account.
+	for _, c := range balance {
+		// Ignore those coins that have a negative value. This
+		// functionality is supposed to be distributing value from
+		// revenue account, not collect it. Otherwise this could be
+		// used to charge the recipients instead of paying them.
+		if !c.IsPositive() {
+			continue
+		}
+
+		// Rest of the division can be ignored, because we transfer
+		// funds to each recipients separately. Any leftover will be
```

Comment:
q: what is the requirement for leftovers? After an "update" message the recipient list may may not contain any of the original recipients. We would need some deterministic way to add it to the recipient with the biggest weight and/or by position for example

Reply:
There is no requirement. 
If I understand correctly what you are proposing, then this is very hard to implement. We would need to keep two states of the `Revenue` so that with every distribution we can check if some of the old recipients should still be paid. And when doing so we are using new recipients money to pay the old ones, so that is not good as well.

If what happens to the leftovers is not that important (edge case?) then I prefer to keep it as simple as possible. Right now it is also easy to explain to humans how it works.

Reply:
I think we can expect a small remainder not to be divisible. But we are talking on something of the order of 10^-6 tokens if a lot. That is small. And can be passed into the next round.

The leftovers coming into a round and those left from the round will average out to the same amount, so this will on average be fair to recipients.